# Final Project: Self Harm and Substance Abuse Deaths Worldwide 

(Last update: 13/12/2023)

Group 1: 
- Thái Mai Khánh Vy - 20120239
- Châu Kiệt -
- Hoàng Tuấn Anh

## In this lab, we're going to be: 
- Find a public data ( https://www.kaggle.com/datasets/thomaseltonau/self-harm-and-substance-abuse-deaths-worldwide )
- Explore data 
- Identify meaningful questions which can be answered with this data
- Preprocess and analyze data to answer each question


## 1. Import libraries

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import plotly.express as px

## 2. Exploring Data

### 2.1 Collecting Data

This data set contains information pertaining to the 'intentional self-harm' and 'psychoactive substance use' related deaths for each country from 2017 to 2021 by sex. Additionally, population data for each country and sex is provided, as well as each countries ISO-3 code.

The authors collected data from multiple sources to compile the dataset on intentional self-harm and psychoactive substance use-related deaths for each country from 2017 to 2021 by sex. Here is a summary of the data collection process: 
- World Health Organisation Mortality Database (2023)
- Kaggle Dataset (2017)
- Our World in Data
- Data Cleaning

After extracting the dataset discovered at the above link, we will have a file named "world_wide_self_harm_and_substance_deaths.csv."

We will proceed to read the data into a DataFrame named "survey_df."

In [9]:
survey_df = pd.read_csv("data/world_wide_self_harm_and_substance_deaths.csv")
survey_df.head()

Year                  Cause Age_Range ISO_Code     Sex  Deaths  \
0  2017  Intentional self-harm         0      GBR  Female       0   
1  2017  Intentional self-harm         0      GBR    Male       0   
2  2017  Intentional self-harm         1      GBR  Female       0   
3  2017  Intentional self-harm         1      GBR    Male       0   
4  2017  Intentional self-harm     10-14      GBR  Female      16   

   Age_and_Sex_Population         Country  
0                743552.0  United Kingdom  
1                782850.0  United Kingdom  
2                768444.0  United Kingdom  
3                810126.0  United Kingdom  
4               3640314.0  United Kingdom

### 2.2 Data exploring & Data preprocessing

Calculate the number of rows and columns for this dataset, these two values are stored in two variables `nrows` and `ncols` respectively.

In [ ]:
nrows, ncols = survey_df.shape

#### Row

The meaning of each row:
Each row in the data set shows Self harm and substance abuse related deaths by country, year (2017-21), and sex

##### Does the data have duplicate rows?

You will test this case and save the results to the `have_duplicated_rows` variable. This variable will have the value True if the data has duplicate lines and will have the value False otherwise.

In [21]:
temp = np.array(survey_df.duplicated())

##Check xem có tồn tại phần tử nào là True hay không, nếu không có bất kỳ một phần tử nào là true thì trả về False
have_duplicated_rows = np.any(temp)
have_duplicated_rows

False

Great, so there are no duplicate rows. Next we will explore the columns.

#### Column

##### The meaning of each column
- Year: Represents the year of the data 
- Cause:  Describes the cause of death. In this dataset, it specifically focuses on "Intentional self-harm" and "Mental and behavioural disorders due to psychoactive substance use"
- Age_Range: Specifies the age range for the reported data. For example, "0" might indicate an aggregate for all ages, while "10-14" and "15-19" and other represent specific age groups.
- ISO_Code: Refers to the ISO code of the country where the data is reported. In the example, "GBR" is the ISO code for the United Kingdom.
- Sex: Indicates the gender for which the data is reported. It could be "Female" or "Male."
- Deaths: Represents the number of deaths for the specified combination of year, cause, age range, country, and gender.
- Age_and_Sex_Population: Denotes the population for the specified age range and gender in the given country for that year.
- Country: Specifies the name of the country for which the data is reported.


##### We will list what these causes of death are. 

List all types of causes and save them in the variable `names_cause`.


In [22]:
names_cause = pd.Series(survey_df["Cause"].unique())
names_cause

0                                Intentional self-harm
1    Mental and behavioural disorders due to psycho...
dtype: object

##### What is the current data type of each column? Are there columns having inappropriate data types?

Get the data type of each column and save it to a series named `cols_dtype`

In [18]:
cols_name = list(survey_df.columns.values)
cols_dtype = pd.Series(survey_df[cols_name].dtypes)
cols_dtype

Year                        int64
Cause                      object
Age_Range                  object
ISO_Code                   object
Sex                        object
Deaths                      int64
Age_and_Sex_Population    float64
Country                    object
dtype: object

##### Data Type Analysis

We have a range of data types, from "Year" to various quantitative data types like death counts and population for the given age, country, and sex, which fall under the integer and numerical data types.

We need to specifically investigate the data types of the "Cause" "Age_Range" "ISO_Code" "Sex" and "Country" columns.

To delve deeper into the data types of these columns, we will write a function. Here, we will use the apply method to perform this analysis.

In [31]:
def open_object_dtype(s):
    dtypes = set()
    
    # We will append each type of s[i] sequentially into the set
    # It is certain that inside dtypes, there will be only one unique type, so this set will contain only one element

    # We extract a series using apply with the type function on all s[i]
    object_series = s.apply(type)
    
    object_series.apply(dtypes.add)
    return dtypes

type_of_cause = open_object_dtype(survey_df['Cause'])
type_of_age_range = open_object_dtype(survey_df['Age_Range'])
type_of_ISO_code = open_object_dtype(survey_df['ISO_Code'])
type_of_age_sex = open_object_dtype(survey_df['Sex'])
type_of_country = open_object_dtype(survey_df['Country'])

##### Comment on Object Data Types

So, we can see that the data types of the columns "Cause" "Age_Range" "ISO_Code" "Sex" and "Country" are strings. Therefore, in this dataset, we will not need to preprocess/normalize the data to perform the required tasks.

##### For each column with numeric datatype, how are the values distributed?

First, we need to see how many missing values the numeric columns have. We need to check these parameters to determine the next course of action, whether we should examine the percentage of missing data and subsequently conduct tests for data distribution. The results will be stored in the variable `missing_values`

In [58]:
numeric_columns = survey_df.select_dtypes(include=['number']).columns
missing_values = survey_df[numeric_columns].isnull().sum()
percentage_missing = (missing_values / len(survey_df)) * 100
summary_statistics = survey_df[numeric_columns].describe().transpose()

In [59]:
print("Missing Values:")
missing_values.head()

Missing Values:


Year                          0
Deaths                        0
Age_and_Sex_Population    19644
dtype: int64

In [60]:
print("Percentage of Missing Values:")
percentage_missing.head()

Percentage of Missing Values:


Year                       0.000000
Deaths                     0.000000
Age_and_Sex_Population    40.393987
dtype: float64

In [61]:
print("Minimum and Maximum Values:")
summary_statistics[['min', 'max']].head()

Minimum and Maximum Values:


min          max
Year                    2017.0       2021.0
Deaths                     0.0      48312.0
Age_and_Sex_Population     0.0  335942016.0

##### For each column with categorical datatype, how are the values distributed?

First, we need to see how many missing values the categorical columns have. We need to check these parameters to determine the next course of action, whether we should examine the percentage of missing data and subsequently conduct tests for data distribution. The results will be stored in the variable `missing_values`

In [49]:
categorical_columns = survey_df.drop(numeric_columns, axis=1).columns
missing_values = survey_df[categorical_columns].isnull().sum()
percentage_missing = (missing_values / len(survey_df)) * 100

print("Missing Values:")
print(missing_values)
print("\nPercentage of Missing Values:")
print(percentage_missing)

Missing Values:
Cause        0
Age_Range    0
ISO_Code     0
Sex          0
Country      0
dtype: int64

Percentage of Missing Values:
Cause        0.0
Age_Range    0.0
ISO_Code     0.0
Sex          0.0
Country      0.0
dtype: float64


Great, so all categorical columns don't have any missing values.

##### How many different values? Show a few

In [62]:
unique_values_info = pd.DataFrame(columns=['Column', 'Number of Unique Values', 'Sample Values'])

for column in categorical_columns:
    num_unique_values = survey_df[column].nunique()
    sample_values = survey_df[column].sample(min(5, num_unique_values)).tolist()
    unique_values_info = pd.concat([unique_values_info, pd.DataFrame({
        'Column': [column],
        'Number of Unique Values': [num_unique_values],
        'Sample Values': [sample_values]
    })], ignore_index=True)

# Display the unique values information
print("Values Information:")
unique_values_info.head()

Values Information:


Column Number of Unique Values  \
0      Cause                       2   
1  Age_Range                      36   
2   ISO_Code                     108   
3        Sex                       4   
4    Country                     109   

                                       Sample Values  
0  [Intentional self-harm, Mental and behavioural...  
1                        [3, 65-69, 35-39, 4, 35-39]  
2                          [URY, GRC, GTM, AUT, MKD]  
3                            [Male, Male, Male, All]  
4   [Paraguay, Luxembourg, Uzbekistan, Oman, Brazil]

#### Data Preprocessing

##### In the dataset, how many countries have been surveyed? Is Vietnam included in this dataset?

We will check the dataset to determine the number of countries and whether Vietnam is present. The results will be stored in two variables: num_countries (an Integer) and isContainsVietnam (a bool).

In [63]:
countries_series = survey_df['Country']
countries_unique = np.unique(countries_series)
num_countries = len(np.unique(countries_unique))

isContainsVietnam = len(countries_unique[countries_unique == 'Vietnam']) != 0

print(num_countries, isContainsVietnam)


109 False


So, the dataset here contains 109 countries, and there is no survey data for the country of Vietnam.

##### The years surveyed in the dataset span from which year to which year?

List the years covered in the dataset, and save the results in three variables: years, min_years, and max_years.



In [64]:
years = np.unique(survey_df['Year'])

min_years = min(years)
max_years = max(years)

print(years,min_years,max_years)

[2017 2018 2019 2020 2021] 2017 2021


Hmmm... Vậy chúng ta có thể thấy rằng, ở đây tất cả các nước đều có đầy đủ dữ liệu từ năm 1990 tới năm 2019